Pulse: Queue Block Generation

Doy-lee opened the pull request:
- Dispatches a worker tagged worker thread in LMQ, "Pulse" that is event driven by Quorumnet and the Blockchain.
- Worker thread sleeps on the Blockchain when it awaits for next round/block to come in if there's no work to do
- Quorumnet is used for communications, and result of said communications is queued into an message queue to be processed by the tagged "Pulse" thread outside of Quorumnet. The responsibility to process the queue is on the caller, i.e. the "Pulse" thread. This is done serially with the benefit of centralising the handling of Pulse state/round progressions into 1 spot and also keeps the thread synchronization simple (mutex + queue) between the quorumnet thread(s) and pulse.

The alternative would be to chain events upon handling a new message but I found this quickly led to sprawly code that jumps from handler functions to more handler functions for each round there'd be.

- Its fairly incomplete, I haven't been able to test against other nodes yet, but sets up the framework for how I think the rest of Pulse will be implemented.

Goes on top of https://github.com/loki-project/loki-core/pull/1190

jagerman commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -2383,6 +2525,9 @@ namespace service_nodes
       x25519_to_pub[derived_x25519_pubkey] = {proof.pubkey, now};
 
+    if (x25519_pkey)
+      *x25519_pkey = derived_x25519_pubkey;
```
See comment elsewhere: I think we should take a `LokiMQ& lmq` in this function and then call update_active_sns() on it from here *if* the ed25519 key changed from what we had stored.

jagerman commented on src/cryptonote_protocol/quorumnet.cpp:
```diff
@@ -88,6 +89,10 @@ struct QnetState {
+    std::mutex pulse_message_queue_mutex;
+    std::condition_variable pulse_message_queue_cv;
+    std::queue<pulse::message> pulse_message_queue;
+
```
See comments in pulse.cpp: basically I think we can avoid all of this by refactoring pulse.cpp to be event-based.  Then instead of storing anything here you would just immediately queue a job for the tagged pulse thread.

jagerman commented on src/cryptonote_protocol/quorumnet.cpp:
```diff
@@ -206,6 +213,13 @@ class peer_info {
+                auto &w = (*qit)->workers;
+                for (size_t i = 0; i < w.size(); i++) {
+                    if (!exclude.count(w[i])) add_peer(w[i]);
+                }
```
The code above needed the index, but this doesn't:
```C++
if (include_workers)
  for (auto &w : (*qit)->workers)
    if (!exclude.count(w) add_peer(w);
```

jagerman commented on src/cryptonote_protocol/quorumnet.cpp:
```diff
@@ -216,12 +230,20 @@ class peer_info {
+                auto &w = (*qit)->workers;
+                for (size_t i = 0; i < w.size(); i++)
+                    add_peer(w[i]);
+            }
```
This looks unnecessary -- the loop added just above looks like it is doing exactly the same thing (except that it also handles `exclude`).

jagerman commented on src/cryptonote_protocol/quorumnet.cpp:
```diff
@@ -1364,6 +1392,233 @@ void handle_blink_success(Message& m) {
+    throw std::runtime_error("Pulse: Daemon is not a service node.");
+
+  pulse::message msg = {};
+  bool in_quorum     = false;
+  for (size_t index = 0; index < quorum.validators.size(); index++)
+  {
+    if (quorum.validators[index] == core.get_service_keys().pub)
+    {
+      msg.quorum_position = index;
+      in_quorum           = true;
+      break;
+    }
+  }
+
+  if (!in_quorum)
+    throw std::runtime_error("Pulse: Could not find this node's public key in quorum");
```
Completely optional suggestion for a more compact way to do it:
```C++
if (auto it = std::find_if(quorum.validators.begin(), quorum.validators.end(),
        [&mypk=core.get_service_keys().pub] (auto const &pk) { return pk == mypk; });
    it == quorum.validators.end())
  throw std::runtime_error("Pulse: Could not find this node's public key in quorum");
else
  msg.quorum_position = it - quorum.validators.begin();
```

jagerman commented on src/cryptonote_protocol/quorumnet.cpp:
```diff
@@ -1364,6 +1392,233 @@ void handle_blink_success(Message& m) {
+
+    assert(validator_bitset != -1);
+    assert(quorum_position != -1);
+    assert(signature);
```
These should be throws rather than asserts since they are triggered by external input which could abort the program (admittedly only for a debug build).  The signature check shouldn't be necessary as the conversion function already throws for a null signature.

jagerman commented on src/cryptonote_protocol/quorumnet.cpp:
```diff
@@ -1364,6 +1392,233 @@ void handle_blink_success(Message& m) {
+    }
+
+    assert(quorum_position != -1);
+    assert(signature);
```
... and here (throw instead of abort)

Doy-lee commented:
Updated with event-based state machine and rebased onto dev. When merging I'm just going to undo the history on the pulse branch so I don't have to repeat rebasing the pulse branch (which I already did for this branch).

jagerman commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -1120,16 +1115,17 @@ namespace cryptonote
-      else
-      {
-        m_lmq->start();
+        m_lmq->add_timer([this]() { pulse::main(m_quorumnet_state, *this); },
```
This `add_timer` can go up with the `add_tagged_thread`; it supports being set up before start() is called (it won't trigger until after startup, of course).

Doy-lee replied:
Updated

jagerman commented on src/cryptonote_protocol/quorumnet.cpp:
```diff
@@ -1659,11 +1657,8 @@ void handle_pulse_participation_bit_or_bitset(Message &m, QnetState& qnet, bool
-    qnet.pulse_message_queue.push(std::move(msg));
-  }
-  qnet.pulse_message_queue_cv.notify_one();
+  auto *self = reinterpret_cast<void *>(&qnet);
```
Conversion to `void*` is implicitly allowed from any pointer type, so you don't need a `reinterpret_cast` here:

```c++
void *self = &qnet;
```
or you could drop it entirely and just capture and reference it in the lambda:
```C++
qnet.lmq.job([&qnet, data = std::move(msg)]() { pulse::handle_message(&qnet, data); }, qnet.core.pulse_thread_id());
```

Doy-lee replied:
Updated

jagerman commented on src/cryptonote_core/pulse.cpp:
```diff
@@ -0,0 +1,724 @@
+      In case of the Service Node network failing, i.e. (pulse round > 255) or
+      insufficient Service Nodes for Pulse, mining is re-activated and accepted
+      as the next block in the blockchain. This resets the genesis block in
+      which future Pulse block timestamps are based off.
```
The difficulty value depends on both the timestamps *and* the difficulty values of the previous N (=60, IIRC) difficulty values, but here we won't have any difficulty values starting at pulse.  I suppose we could just fix it at something for the first couple blocks and let the DA build it from there.
